### ISSUE TITLE
Support additional level of nesting?

### ISSUE THREAD

**Issue by tlambert03** [napari/napari.github.io#308]
jupyter-book supports multiple levels of nesting in the top level TOC.  As I understand from @codemonkey800, we're currently restricted to a single level: 

<img width="271" alt="Untitled" src="https://user-images.githubusercontent.com/1609449/146471018-298cae8a-8236-4bf1-8a02-a689401818a5.png">

would be nice to lift that restriction :)

**Comment by tlambert03** [comment]
As I look at updating the plugin documentation, I keep coming up against this limitation.  The lack of additional level of nesting here means that everything in plugins must be a top level document, and all nesting must essentially happen in-page.  That leads to very long overwhelming documents.  So just want to poke this issue a bit and make sure it does get some discussion

**Comment by codemonkey800** [comment]
@liaprins-czi any thoughts? i know that we designed the UI to not have multiple levels of nesting on the left hand sidebar 🤔

**Comment by DragaDoncila** [comment]
On this topic I'm a bit conflicted - I do think it's important to view navigation on the left "at a glance". It's frustrating when you expand a left-side nav and it takes up the whole viewport.

I do think it would be nice to have an additional level of nesting on the left (sort of like subfolders within the broad folders). And, I also think we should nest down to `###` on the right-side nav for within pages. There's been a few times where I've had to rearrange docs *just* so I could make sections have `##` headers so they could show up on the right.

Just to clarify, I mean this nav.

![image](https://user-images.githubusercontent.com/17995243/146614214-d14467cf-a535-479a-8b69-de885b8c25fe.png)

**Comment by tlambert03** [comment]
> There's been a few times where I've had to rearrange docs just so I could make sections have ## headers so they could show up on the right.

yeah, I see that that's basically the only workaround... and currently, I'm frustrated by how _long_ it makes various pages.  I think there's a legitimate cognitive burden when you open a page, scroll quickly to see how long it is, and think "omg!  understanding this is going to take forever!"

> On this topic I'm a bit conflicted - I do think it's important to view navigation on the left "at a glance". It's frustrating when you expand a left-side nav and it takes up the whole viewport.

nesting should help, not hurt this?  (all sub-levels should not be shown at once, but each collapsible, as they were on the Fura theme, or on the standard jupyter book theme).

**Comment by DragaDoncila** [comment]
> nesting should help, not hurt this? (all sub-levels should not be shown at once, but each collapsible, as they were on the Fura theme, or on the standard jupyter book theme).

Oh yeah lol. That would be good. I do remember some concerns with our previous documentation being *too* nested, but I think we should control that through our docs structure rather than through no nesting.

**Comment by tlambert03** [comment]
> I do remember some concerns with our previous documentation being too nested

Yeah, I guess the problem now is that, we use the first level of nesting to group _major_ things.  So, once we get into a topic (like plugins) we're basically out of levels.  I'd really love to be able to organize our plugin docs a bit more like this: https://code.visualstudio.com/api  (specifically: "get started", "guides", "legacy engine" and "schema & reference" sections)... they don't look great all flat and next to each other, but that's the only option

**Comment by sofroniewn** [comment]
> Yeah, I guess the problem now is that, we use the first level of nesting to group major things. So, once we get into a topic (like plugins) we're basically out of levels.

Yeah, i think this is a really good point. We also don't really have nesting at the level of our API reference, as the high-level stuff is always there.

For example on numpy docs

![image](https://user-images.githubusercontent.com/6531703/148142563-56e2a8e8-6788-48d9-a87a-bc9013903638.png)

you can see the nesting around their different routines here

If I look at the layers in the napari docs I don't get any further nesting as `napari.layers` takes it all up

![image](https://user-images.githubusercontent.com/6531703/148142616-5d9425c8-2e8e-4e3e-bbdb-1fa69c21d396.png)

I'd really expect here to see the different layers `Image`, `Labels` etc and be able to move rapidly between them.

I know @neuromusic is planning for some improvements to the docs over the next couple weeks - maybe we can get this on the docket as a high priority item as it should improve the overall docs usability. cc @justinelarsen @liaprins-czi

**Comment by justinelarsen** [comment]
I prefer having the navigation expanded on the left with multiple layers of nesting as well. This improvement is on our roadmap to work on in the next couple of weeks. It looks like it's related to this issue https://github.com/napari/napari.github.io/issues/275

**Comment by neuromusic** [comment]
> Yeah, I guess the problem now is that, we use the first level of nesting to group major things. So, once we get into a topic (like plugins) we're basically out of levels.

This resonates with me... the current _design_ is consistent with the VS code docs (2-level hierarchy on the left, second levels are pages, within-page nav on the right). VS code then maintains the top-level sections as navigation in the top bar.

Being more consistent with VS code would require (a) some content reorganization to organize the docs into <6 or so top-level "sections" where each would have its own unique left nav (cc @melissawm) and (b) some design work to add a new top-level nav in the header or somewhere (cc @liaprins-czi).

**Comment by tlambert03** [comment]
I'm not super concerned with being consistent with VS code.  Obviously, they split their plugin docs into a completely different top level URL from their end-user docs.  I assume people don't want to break our plugin docs into a new top level page.  In light of that, maintaining the natural nesting provided by the other sphinx-compatible themes like fura, pydata, etc. seems like the most direct path here

**Comment by neuromusic** [comment]
> I'm not super concerned with being consistent with VS code.

Oh, agreed... I'm just looking at them and numpy etc for inspiration on documentation structuring.

> maintaining the natural nesting provided by the other sphinx-compatible themes like fura, pydata, etc. seems like the most direct path here

Yes, the implementation should take a basic sphinx-structured ToC and map items accordingly, but what both the numpy and VS code examples would point to is that the topmost level of the sphinx ToC should not be in the left sidebar, but elsewhere (header?) and the left nav should be the second and third levels of the hierarchy. Any deeper hierarchy than that should be within-page and navigable through the right nav.

That would give us the ability to support 3 levels of **page** hierarchy rather than the current 2.

**Comment by liaprins-czi** [comment]
> What both the numpy and VS code examples would point to is that the topmost level of the sphinx ToC should not be in the left sidebar, but elsewhere (header?) and the left nav should be the second and third levels of the hierarchy. 

I am seeing the underlying problem here as being able to navigate to a third level of hierarchy in some way, but I think it would not necessarily need to be the way numpy and / or VS Code do it ("with topmost level of the not be in the left sidebar, but elsewhere") ... I think we can keep the topmost level in the left sidebar if there is a design that works (and this would be better for narrower screen sizes wherein the left nav (and topmost nav) cannot always fit on screen by default and have to be in a hamburger / hidden menu.

This issue came up a while ago I think it was fairly soon after the napari.org bug bash, to be able to add one more level of hierarchy for the API page specifically, and I did some explorations here:
https://www.figma.com/file/CIgWTb4iTAcpVj5JpwuSJU/?node-id=0%3A1

I remember discussing it with @codemonkey800 at the time and so I thought we had already implemented an [update](https://www.figma.com/file/CIgWTb4iTAcpVj5JpwuSJU/Solution-explorations?node-id=2%3A537) to accommodate the extra level of hierarchy... but would this fix the issues raised? To me it seems like it does, but what do others think?

I should be able to work on this more in the coming sprint, if needed!

**Comment by codemonkey800** [comment]
We currently only have "nesting" on one page, but it's not on the left hand nav. Instead, it's on the page directly:

<img width="300" src="https://user-images.githubusercontent.com/2176050/148268641-1a314d2f-4440-487e-8f4b-4e39dc52fbcf.png">

This was implemented in #192 using these [designs](https://www.figma.com/file/Vm8z7DSjDjJurEX1RxEA39/Styling?node-id=928%3A2), but I can't seem to find the discussion related to it.

Either way, we can definitely implement [any one of the nesting designs](https://www.figma.com/file/CIgWTb4iTAcpVj5JpwuSJU/Solution-explorations?node-id=2%3A537). I think the indented list makes the most sense to me, but we would need to be restrictive on the number of levels supported so that we don't have long overflowing text in the sidebar

**Comment by melissawm** [comment]
I am +1 on 3 levels of nesting. I 100% agree with @tlambert03 about the cognitive load of long documents, and one more level should be enough for that, at least in light of the current docs.

**Comment by liaprins-czi** [comment]
I have made a prototype showing the design for this third level of nav nesting for our docs. I made of the names of some imaginary docs at the deepest level, FYI! Also the top-level sections shown in the prototype are not necessarily the same as the latest structure proposed by @melissawm ! But this should serve to get the concept across, and show interaction and behavior in various situations:
[Link to Figma prototype](https://www.figma.com/proto/CIgWTb4iTAcpVj5JpwuSJU/Deeper-navigation-hierarchy?page-id=133%3A2&node-id=148%3A115&viewport=241%2C48%2C0.5&scaling=scale-down&starting-point-node-id=148%3A115)

I also detailed the interaction and behavior for all the edge cases and other situations I could think of in this G-Doc:
[Link to doc of considerations](https://docs.google.com/document/d/1oarCoerYSOogDp2kv7b3Q6sD-GvqfkxftUwTtsvw7DA/edit?usp=sharing)

@codemonkey800 this design is pending implementation feasibility... if you are able to tell if its various aspects / interaction proposals are possible or not now, then I can work on updating parts that wouldn't work. We can also iterate later if something comes up while you are implementing.nThanks!

We all still need to align on sometihng tangential to and that will be brought about by adding a third level of nesting:
[**What will the content be for the _pages_ that have children, but are not the main section’s ToC page?**](https://docs.google.com/document/d/1oarCoerYSOogDp2kv7b3Q6sD-GvqfkxftUwTtsvw7DA/edit#heading=h.x4me0l6fvj12)

- Will they have their own content, like any other at its same level without children (i.e. see “npe2 migration guide” in left sidebar for example)?
- If so, will it also have links to its children pages?
- If so, where should they live (at top or bottom of page)?
- Or will they be mini-ToC pages, just listing out their children as bulleted links, same idea as “Plugins” page does?
- Or should we make it so users cannot navigate to them, only click the icon to expand their children in the left sidebar?
Whatever we do, it needs to be consistent for all of these types of pages, so we need to align on the content strategy for this.
@neuromusic Do you think this ^^^ is something for Documentation Working Group to consider?

Thanks all!

**Comment by liaprins-czi** [comment]
@DragaDoncila re your earlier comment about the _in-page_ ToCs on right side margin:
> I also think we should nest down to ### on the right-side nav for within pages

Just to verify, this would mean just adding one more level of nesting in that area, right?

I can work on this design as well soon. However, I think it is independent from the design + implementation work for the _left_-side nav (@codemonkey800 do you agree?). Should it have a separate GH issue to track it?

**Comment by codemonkey800** [comment]
@liaprins-czi thank you Lia!

> this design is pending implementation feasibility... 

It looks feasible engineering wise! A new icon and some CSS to add left margin is all we need. 

>  Should it have a separate GH issue to track it?

I think so, this issue is specifically for the left hand nav.

**Comment by liaprins-czi** [comment]
@codemonkey800 Okay great, here is a link to the icon in Figma:
https://www.figma.com/file/pDKU0CdSzpSWGaoXXQVJLz/v0-hub-final-designs?node-id=1342%3A98

And I can create a separate issue for just the right-nav now.
